New  mask above/below threshold preprocessors

ledm opened the pull request:
Hi, 

I have a few changes I'd like to pull into the main branch related to the the `mask_threshold` preprocessor in `esmvaltool/preprocessor/_mask.py`.

There is already a threshold masking preprocessor, but:
1. `mask_threshold` doesn't appear in the `__all__` list of preprocessors. So we can't use it anyway. 
2. `mask_threshold` only masks below the threshold.
3. We are sometimes interested in a range of values , so both a minimum and maximum threshold would be needed.

The solution was to create two new  preprocessors in _mask: `mask_below_threshold` and `mask_above_threshold`. I'm not sure if `mask_threshold`  was being used elsewhere though, so  I copied it and renamed it `mask_below_threshold`. 

Please have a look at the changes below, and let me know if everything is okay. 

Note that this PR is not associated with any numbered issue.

Lee

bouweandela commented:
I'm pretty sure the old function was not used by anyone, because as you say, it was not possible to use it, so I think it can be safely deleted.

Wouldn't it make more sense to provide the functions [`masked_inside`](
https://docs.scipy.org/doc/numpy/reference/generated/numpy.ma.masked_inside.html#numpy.ma.masked_inside) and [`masked_outside`](
https://docs.scipy.org/doc/numpy/reference/generated/numpy.ma.masked_outside.html#numpy.ma.masked_outside)
instead of the ones you have chosen now?

valeriupredoi commented:
yeah @ledm Lee man, ```mask_threshold``` is indeed one of those relics from my very first sets of implementations from eons ago -- delete it or rename it whatsoever -- and yes, probably best to use ```mask_where``` or as @bouweandela says inside and outside, this way we don't have to memcopy the cube

ledm commented:
Hi, I deleted the `mask_threshold` preprocessor, removed the parts where the arrays were copied, and added the masked_inside_range and mask_outside_range preprocessors (for completeness).

valeriupredoi commented:
@ledm good stuff, man! Two things tho: you can get rid of the import of ma by using np.ma instead (numpy is already imported at top level), the second thing is ```ma.masked_where(mycube.data > threshold, mycube.data)``` will mask out all elements that are greater than your ```threshold```, and conversely the other way round for the complementary function -- justt checking with you this is what you need since the docstrings say the opposites

ledm commented:
Removed the numpy imports, and changed the doc string. Cheers!

ledm commented:
Hi @valeriupredoi, @bouweandela,

Just so you know, I believe that this PR is ready to go. If there's anything else needed to get it authorised, please let me know as soon as you can as I'm heading out for holidays today from noon. I will not be able to contribute anything after the next three hours if any further changes are required. 

Also, I'd like to avoid another long an drawn out PR. #438 was a bit of a nightmare. Cheers!

Lee

valeriupredoi commented:
PR looks nice and shprt, will merge after @mattiarighi has tested (or fix
anything and merge). Have fun, bud!

On Wed, 1 Aug 2018, 09:22 Lee, <notifications@github.com> wrote:

> Okay, thanks @mattiarighi <https://github.com/mattiarighi>
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/ESMValGroup/ESMValTool/pull/549#issuecomment-409492501>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AbpCo5iNlAr3hDe9VdhWqYktvoIeMOUUks5uMWU2gaJpZM4VoW7j>
> .
>

valeriupredoi commented:
am gonna delete the branch as well

valeriupredoi commented:
ahh @mattiarighi you beat me :grin:

mattiarighi commented:
:zap:

bouweandela commented:
This was merged a bit too quickly: the documentation misses Arguments and Returns sections and the recipe should probably be moved to the examples directory.
